### PR TITLE
Send log messages to the constellation

### LINKED ISSUE #11776: Include warnings in crash reports
We should buffer up `warn!`, and send an error report on `error!` as well as panic. (We need to make sure that privacy-sensitive data doesn't end up in crash reports, possibly using types/traits to track privacy.)

cc @paulrouget

### PR DESCRIPTION
**Pull request by asajeffrey**
Send all warnings and errors to the constellation. Warnings are bufferred up, and included in any subsequent error reports. Errors are reported in the same way as panics.

Note that this can't merge yet, as it needs https://github.com/rust-lang-nursery/log/pull/86 to land.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11776 (github issue number if applicable).
- [X] These changes do not require tests because we don't test crash reporting.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11841)

### CODE REVIEW

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
https://github.com/rust-lang-nursery/log/pull/86 has landed, it just needs to make its way to crates.io.

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
env_logger 0.3.4 is now out, so this can merge. r? @Ms2ger

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -304,6 +312,97 @@ enum ExitPipelineMode {
     Force,
 }
 
+/// A logger directed at the constellation from content processes
+#[derive(Clone)]
+pub struct FromScriptLogger {
```
Comment (Manishearth) [inline components/constellation/constellation.rs]: We could probably have a `FromLogger<T>` struct with a common Log impl, where `T` is `FromFooMsg`, and implements a trait which provides a `get_msg(Option<PipelineId>, Option<String>, LogEntry) -> Self` method

**Comment by Manishearth** [comment]
r=me with or without that trait abstraction

**Comment by asajeffrey** [comment]
I think since there's only two instances, the extra level of generality won't buy us that much. Thanks for looking the code over!

@bors-servo r=Manishearth

**Comment by asajeffrey** [comment]
OK, that's weird, did I accidentally fix a bug?

**Comment by asajeffrey** [comment]
IRC conversation: http://logs.glob.uno/?c=mozilla%23servo&s=15+Jul+2016&e=15+Jul+2016#c480756

TL;DR: when logs are reported to the constellation, we can check to see if the thread has panicked, and report `Error`s as `Panic`s. The issue reporting and hard-failing mechanism can then just be used for panic, in particular hard-fail won't be triggered by JS errors.

**Comment by asajeffrey** [comment]
@Manishearth care to look over that last commit?

**Comment by asajeffrey** [comment]
Rebased.

@bors-servo r=Manishearth (http://logs.glob.uno/?c=mozilla%23servo&s=15+Jul+2016&e=15+Jul+2016#c480930)

**Comment by asajeffrey** [comment]
@bors-servo retry
 * CI oddness
